### PR TITLE
systemd: type some card components

### PR DESCRIPTION
**Pull request by martinpitt**
This found a bug where we were previously trying to get non-existing "from" and "port" field from the getLastlog2() result.

Plus some lib updates to support this.

### CODE REVIEW

File: pkg/systemd/overview-cards/lastLogin.tsx [hunk 1]
```diff
@@ -84,12 +91,12 @@
         }
     });
 
-    const lastLoginTime = messages?.['last-login-time'] ?? lastlog?.time;
+    const lastLoginTime: number | undefined = messages?.['last-login-time'] ?? lastlog?.time;
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/lastLogin.tsx:94]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/lastLogin.tsx [hunk 2]
```diff
@@ -84,12 +91,12 @@
...
     if (!lastLoginTime)
         return null;
 
-    const lastLoginFrom = messages?.['last-login-host'] || lastlog?.from;
-    const lastLoginPort = messages?.['last-login-line'] || lastlog?.port;
+    const lastLoginFrom = messages?.['last-login-host'] || lastlog?.host;
+    const lastLoginPort = messages?.['last-login-line'] || lastlog?.tty;
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/lastLogin.tsx:99]: These 2 added lines are not executed by any test.

File: pkg/systemd/overview-cards/lastLogin.tsx [hunk 3]
```diff
@@ -53,10 +53,17 @@ const getFormattedDateTime = (time) => {
     return timeformat.dateTime(date);
 };
 
+type LoginMessages = {
+    "last-login-time": number,
+    "last-login-host": string,
+    "last-login-line": string,
+    "fail-count"?: number,
```
Comment (jelly) [inline pkg/systemd/overview-cards/lastLogin.tsx:60]: Why can this be undefined? In cockpit session it is initialized as 0

File: pkg/systemd/overview-cards/lastLogin.tsx [hunk 4]
```diff
@@ -65,7 +72,7 @@ const LastLogin = () => {
         const bridge = cockpit.dbus(null, { bus: "internal" });
         try {
             const [reply] = await bridge.call("/LoginMessages", "cockpit.LoginMessages", "Get", []);
-            const obj = JSON.parse(reply);
+            const obj = JSON.parse(reply as string[0]);
```
Comment (jelly) [inline pkg/systemd/overview-cards/lastLogin.tsx:75]: What is `as string[0]`?

File: pkg/systemd/overview-cards/lastLogin.tsx [hunk 5]
```diff
@@ -84,12 +91,12 @@ const LastLogin = () => {
         }
     });
 
-    const lastLoginTime = messages?.['last-login-time'] ?? lastlog?.time;
+    const lastLoginTime: number | undefined = messages?.['last-login-time'] ?? lastlog?.time;
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/lastLogin.tsx:94]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/lastLogin.tsx [hunk 6]
```diff
@@ -84,12 +91,12 @@ const LastLogin = () => {
...
     if (!lastLoginTime)
         return null;
 
-    const lastLoginFrom = messages?.['last-login-host'] || lastlog?.from;
-    const lastLoginPort = messages?.['last-login-line'] || lastlog?.port;
+    const lastLoginFrom = messages?.['last-login-host'] || lastlog?.host;
+    const lastLoginPort = messages?.['last-login-line'] || lastlog?.tty;
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/lastLogin.tsx:99]: These 2 added lines are not executed by any test.

File: pkg/systemd/overview-cards/shutdownStatus.tsx [hunk 7]
```diff
@@ -72,13 +73,15 @@ export const ShutDownStatus = () => {
...
-        displayDate = timeformat.dateTimeNoYear(date);
-    } else {
-        displayDate = timeformat.dateTime(date);
+    if (shutdownTime) {
+        const date = new Date(shutdownTime / 1000);
+        const now = new Date();
+        if (date.getFullYear() == now.getFullYear()) {
+            displayDate = timeformat.dateTimeNoYear(date);
+        } else {
+            displayDate = timeformat.dateTime(date);
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/shutdownStatus.tsx:83]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/lastLogin.tsx (hunk 6 above)
Comment (jelly) [inline pkg/systemd/overview-cards/lastLogin.tsx:99]: This was never tested and always wrong?

File: pkg/systemd/overview-cards/lastLogin.tsx (hunk 3 above)
Comment (martinpitt) [reply pkg/systemd/overview-cards/lastLogin.tsx:60]: No, it doesn't get written if fail_count == 0, see https://github.com/cockpit-project/cockpit/blob/main/src/session/session-utils.c#L338

File: pkg/systemd/overview-cards/lastLogin.tsx (hunk 4 above)
Comment (martinpitt) [reply pkg/systemd/overview-cards/lastLogin.tsx:75]: Ah, that's bogus, leftover from  some earlier variant. Thanks for spotting! Fixed to `as string`. The return value of `.call()` is currently `unknown[]`, I don't think we can restrict it much further. So we have to make explicit casts on method return values based on what we know about their signature.

File: pkg/systemd/overview-cards/lastLogin.tsx (hunk 6 above)
Comment (martinpitt) [reply pkg/systemd/overview-cards/lastLogin.tsx:99]: Yes :cry:

File: pkg/systemd/overview-cards/lastLogin.tsx (hunk 5 above)
Comment (cockpituous) [inline pkg/systemd/overview-cards/lastLogin.tsx:94]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/lastLogin.tsx (hunk 6 above)
Comment (cockpituous) [inline pkg/systemd/overview-cards/lastLogin.tsx:99]: These 2 added lines are not executed by any test.

File: pkg/systemd/overview-cards/shutdownStatus.tsx (hunk 7 above)
Comment (cockpituous) [inline pkg/systemd/overview-cards/shutdownStatus.tsx:83]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/lastLogin.tsx (hunk 3 above)
Comment (jelly) [reply pkg/systemd/overview-cards/lastLogin.tsx:60]: But we only print when we call `cockpit_json_print_integer_property`? Otherwise we `goto out`.

**Comment by martinpitt** [comment]
@jelly PTAL?

**Review by jelly — APPROVED** [review]

**Pull request merged**